# MMP3C framework for the charaterization of tumor metabolic plasticity
## Description
Metabolic reprogramming, as one of hallmarks in cancer cells, enables to meet divergent demands for energy, biomass to maintain tumorigenesis, metastasis and drug resistant. A typical example refers to Warburg effect, preference of glycolysis as energy source rather than oxidized phosphorylation (OXPHOS) in spite of sufficient oxygen presence. This process requires its metabolic plasticity and flexibility. However, given metabolic intratumour heterogeneity raised by different cancer types, landscape analysis of tumor metabolic plasticity spanning different cancer types, in particular, metabolic crosstalk within in distinct cell subpopulations surround tumor microenvironment, is scarce. In this perspective, we aim to cover this gap by developing a new in-silico framework, termed as MMP3C (Modeling Metabolic Plasticity by Pathway Pairwise Comparison), to depict tumor metabolic plasticity based on transcriptome.

## MMP3C framework
Source codes for generating results of "MMP3C: An in-silico framework to depict cancer metabolic plasticity using gene expression profiles"

## Installation
All packages can be installed via bioconductor (https://www.bioconductor.org/), GitHub, and CRAN. Generally, a couple of minutes is needed for installing each package.
